Guide/file-storage.markdown

```diff
@@ -575,3 +575,244 @@ config = do
+
+```haskell
+[hsx|<img src={pathTo $ RenderImageStyleAction 400 200 imageUrl} />|]
+where
+    imageUrl = "http://localhost:8000/static/picture/b4c8f55c-16d6-41f0-9503-77352b134e14"
+```
+
+### Security Considerations
+
+The above implementation is not secure. Anyone can request any image style of any image. This may put a burden on the site. Imagine a malicious user requesting thousands of different dimensions of a 100MB image. The server would have to generate thousands of images. This would put a lot of load on the server. To prevent this, we need to add some access control. The way we would do it, as that is the simplest way, is to add a signed token to the image style URL. The token would be generated by the server and would contain the original image path, the desired width and height. The server would then verify the token and only serve the image style if the token is valid. This way, the server would not have to generate the image style on the fly, but only if the token is valid. If the token is not valid, then the server would return an access denied.
+
+We start by adding the `cryptonite` package to the `default.nix` file, and re-start `./start` to get the package.
+
+```nix
+# default.nix
+haskellEnv = import "${ihp}/NixSupport/default.nix" {
+    ihp = ihp;
+    haskellDeps = p: with p; [
+        cabal-install
+        base
+        # ...
+        cryptonite # <--- ADD THIS
+    ];
+```
+
+Then we need to generate a private and public key pair. We can do it in our Config file, and save the keys for later use.
+
+```haskell
+-- Config/Config.hs
+import Crypto.PubKey.RSA as RSA
+
+newtype RsaPublicAndPrivateKeys = RsaPublicAndPrivateKeys (RSA.PublicKey, RSA.PrivateKey)
+
+config :: ConfigBuilder
+config = do
+    -- See https://ihp.digitallyinduced.com/Guide/config.html
+    -- for what you can do here
+
+    -- ...
+
+    -- Private and public keys to sign and verify image style URLs.
+    -- The 300 means how large your URL can be.
+    -- The 65537 is the public exponent of RSA.
+    -- You can probably leave those values as is.
+    (publicKey, privateKey) <- liftIO $ liftIO $ RSA.generate 300 65537
+    option $ RsaPublicAndPrivateKeys (publicKey, privateKey)
+```
+
+Your Controller will now look like this:
+
+```haskell
+module Web.Controller.ImageStyle where
+
+import Web.Controller.Prelude
+import IHP.ControllerSupport
+import System.Directory (doesFileExist)
+import qualified Data.Text as T
+
+-- Imports for the signed token.
+import Crypto.PubKey.RSA.PKCS15 as RSA
+import Crypto.Hash.Algorithms as Hash.Algorithms
+import Config
+import Data.ByteString.Base64 as Base64
+
+instance Controller ImageStyleController where
+    action RenderImageStyleAction { width, height, originalImagePath, signed } = do
+        let size = show width <> "x" <> show height
+
+        -- Verify the signed token.
+        let Config.RsaPublicAndPrivateKeys (publicKey, _) = getAppConfig @Config.RsaPublicAndPrivateKeys
```

Comment:
This can also be moved to a top level declaration:

```suggestion
publicKey :: (?context :: ControllerContext) => PublicKey
publicKey = (getAppConfig @Config.RsaPublicAndPrivateKeys).publicKey
```